Linter structure changes + extra chartfile rules

migmartri opened the pull request:
This is still WIP but I'd like to get some feedback in case you guys think I am going in the wrong direction.

Currently the PR includes:

Structural changes:
- To show "Lint OK" message if everything went well. I thought that this is nice so you the user gets some feedback.
- Linter package split in two more sub-packages, once for rules and another for `support structures/functions`. I do not like the name support so feedback is more than welcome :)
-  New linter type that works in a similar fashion than `validator` in helm classic.
- linter.RunLinterRule(SEVERITY, VALIDATION_FUNCTION), which is in charge of populate the messages slice.
- validationFunction enforcement. I wanted to try to enforce the usage of validation functions to composite the linter itself. A validation function is one that returns a LintError. These functions should be small enough that could be easily tested.  
- linter.RunLinterRule and its validationFunction can and should be run in isolation, making easier to add and maintain extra rules. In any case, It is possible to check the execution of a previous runLinterRule in case that we want to stop the linter execution (see validChartFile in Charfile()). 
- Moved the current lint rules for templates and values to the new format. Those rules will likely change once we work on the extra rules.

New  lint rules:
- [x] Chart name of the chart and the directory name are the same.
- [x] Validate that the version is bigger than 0.0.0 using semver comparison instead of string comparison.
- [x] Validate that it contains a valid object type, for example
- [x] Engine is one of the valid values.
- [x] Each maintainer contains a name
- [x] Email has email format
- [x] Source and home are URL’s (if provided)

I will create two more PR with the new rules for templates and values.yaml.

I will also create a new one with a test suite for the new validation functions.

Commit:
New Chart linter structure

Fixed tests

Adding chart name linter

Add lint error

Moving to blocks

Moving to method

Moved lint rules to functions

Semantic version validation

Linting engine

Adding sources and home validations

Sharing file loading

Sharing file loading

Rolling back readme

Rewriting other linters

Fixing tests

Typo

Using chart.Metadata

Fixing format

Adding UNKNOWN in Engine

Adding tabs

Fixing tabs

migmartri commented:
It seems that some of the changes done in `chart` package are not needed. Actually that package should not be used. Use chartutil and pkg/proto/hapi/chart

https://kubernetes.slack.com/archives/helm/p1466007337000070
https://kubernetes.slack.com/archives/helm/p1466007374000071

michelleN commented on pkg/lint/rules/chartfile.go:
```diff
@@ -0,0 +1,140 @@
+	chartFile, err := chart.LoadChartfile(chartPath)
+	validChartFile := linter.RunLinterRule(support.ErrorSev, validateChartYamlFormat(linter, err))
+
+	// Guard cause. Following linter rules require a parseable ChartFile
```
👍

michelleN commented:
This looks very good so far @migmartri. 
I'll let you know if I think of something other than "support" but honestly, I don't have an issue with that either.

Were you thinking we could merge this PR and then have another PR for the rest of the rules or did you want to leave this one as a WIP?

migmartri commented:
> Were you thinking we could merge this PR and then have another PR for the rest of the rules or did you want to leave this one as a WIP?

After your feedback I plan on fixing a couple of issues on this one like using chartutil instead of chart and then ask for merging it. I'll let you know :)

I will create one or two more PR's with the rest of the rules which I hope to have them available by next week.

migmartri commented:
Just updated the PR.

The main changes are:

* Use chartutils instead of legacy chart package.
* Updated chart.Metadata.proto to include an enum field called Engine. So there is a place where we list the valid values for the Engine and can use that generated exported var in the linter. `validateChartEngine`

Once closed this one, I will start working on the lint rules for values and templates.

fibonacci1729 commented on _proto/hapi/chart/metadata.proto:
```diff
@@ -17,6 +17,10 @@ message Maintainer {
 message Metadata {
+	enum Engine {
+		// gotpl is the default engine
+		gotpl = 0;
```
style: change 'gotpl' to 'GOTPL' (proto uses capitals with underscores for enum values.). Also, using 0 to represent the default could be dangerous, because unsupported values will default to GOTPL, i.e. making it hard to distinguish between someone supplying a default vs. supplying an incorrect value.

migmartri replied:
Awesome feedback, I will work on it.

The thing is that I am not sure if I should use an Enum here, because as you can see, I am not using it as the type for Engine in the Metadata Message. I am not using it in the Engine type because then the yaml parser complains because the the imported value is a string not an int32 (expected by enum)

So in the end I use the Enum just to store the different options and load them in the linter. 

Do you have any other suggestion on how to achieve my goal?

technosophos replied:
I added the Engine type just a few weeks ago, and I think it's only used in one place in the code. I would be happy to change that to use the Enum you supplied here. We reserved the right to  change the protos (in a backward compatibility breaking way) up until Beta 1, so we could do that if it's a better fit.

fibonacci1729 commented:
lgtm
